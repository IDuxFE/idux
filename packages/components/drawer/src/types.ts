import type { DefineComponent } from 'vue'

import { IxExtractPropTypes, IxPropTypes } from '@idux/cdk/utils'

type Done = (cancel: boolean) => void
export type BeforeCloseFunc = (done: Done) => void

export const drawerProps = {
  visible: IxPropTypes.bool,
  title: IxPropTypes.string,
  footer: IxPropTypes.string,
  closable: IxPropTypes.bool.def(true),
  placement: IxPropTypes.oneOf(['top', 'right', 'bottom', 'left'] as const).def('right'),
  width: IxPropTypes.oneOfType([Number, String]),
  height: IxPropTypes.oneOfType([Number, String]),
  offset: IxPropTypes.oneOfType([Number, String]).def(0),
  mask: IxPropTypes.bool.def(true),
  maskClosable: IxPropTypes.bool.def(true),
  wrapClassName: IxPropTypes.string,
  destroyOnHide: IxPropTypes.bool.def(false),
  keyboard: IxPropTypes.bool.def(true),
  beforeClose: IxPropTypes.func<BeforeCloseFunc>(),
}

export type DrawerProps = IxExtractPropTypes<typeof drawerProps>

export type DrawerInstance = InstanceType<DefineComponent<DrawerProps>>
