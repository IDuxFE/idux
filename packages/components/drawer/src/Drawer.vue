<template>
  <ix-portal target="body">
    <transition name="ix-drawer-fade" @after-enter="afterEnter" @after-leave="afterLeave" @before-leave="beforeLeave">
      <ix-mask v-show="maskVisible" :mask="mask" :mask-class="wrapClassName" @click="onModalClick">
        <div ref="drawerRef" :class="['ix-drawer', placement]" :style="drawerStyle" @click.stop>
          <header v-if="title || slots.title" class="ix-drawer-header">
            <slot name="title">
              <span role="heading" :title="title">
                {{ title }}
              </span>
            </slot>
            <button v-if="closable" class="ix-drawer-close-btn" type="button" @click="handleClose">
              <ix-icon class="ix-icon-close" name="close" />
            </button>
          </header>
          <section v-if="rendered" class="ix-drawer-body">
            <slot></slot>
          </section>
          <footer v-if="footer || slots.footer" class="ix-drawer-footer">
            <slot name="footer">
              <span role="footer" :title="footer">
                {{ footer }}
              </span>
            </slot>
          </footer>
        </div>
      </ix-mask>
    </transition>
  </ix-portal>
</template>
<script lang="ts">
import { defineComponent, ref, computed } from 'vue'
import type { SetupContext } from 'vue'
import { drawerProps, DrawerProps } from './types'
import { IxPortal } from '@idux/cdk/portal'
import { IxIcon } from '@idux/components/icon'
import { IxMask } from '@idux/components/mask'
import useModal from './useModal'

export default defineComponent({
  name: 'IxDrawer',
  components: { IxPortal, IxIcon, IxMask },
  props: drawerProps,
  emits: ['open', 'opened', 'close', 'closed', 'update:visible'],
  setup(props: DrawerProps, ctx) {
    const drawerRef = ref<HTMLElement>(null)
    const isHorizontal = computed(() => props.placement === 'right' || props.placement === 'left')
    const renderPx = (size?: string | number) => {
      return typeof size === 'number' ? `${size}px` : size
    }
    const drawerStyle = computed(() => {
      const width = props.width || (isHorizontal.value ? '30%' : '100%') // 不传值时水平方向宽度默认30%
      const height = props.height || (isHorizontal.value ? '100%' : '30%') // 不传值时垂直方向高度默认30%
      const renderOffset = renderPx(props.offset)
      const offset = isHorizontal.value ? { top: renderOffset } : { left: renderOffset }
      return {
        width: renderPx(width),
        height: renderPx(height),
        ...offset,
      }
    })
    return {
      ...useModal(props, ctx as SetupContext, drawerRef),
      slots: ctx.slots,
      drawerStyle,
    }
  },
})
</script>
