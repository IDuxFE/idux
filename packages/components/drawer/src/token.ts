/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { DrawerBindings, DrawerProps } from './types'
import type { DrawerConfig } from '@idux/components/config'
import type { InjectionKey, Ref, Slots } from 'vue'

export interface DrawerContext {
  props: DrawerProps
  slots: Slots
  config: DrawerConfig
  visible: Ref<boolean>
  animatedVisible: Ref<boolean>
}

export const drawerToken: InjectionKey<DrawerContext> = Symbol('drawerToken')

// public token
export const DRAWER_TOKEN: InjectionKey<DrawerBindings & { props: DrawerProps }> = Symbol('DRAWER_TOKEN')
