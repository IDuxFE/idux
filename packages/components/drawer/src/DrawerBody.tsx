/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { Slot } from 'vue'

import { computed, defineComponent, inject } from 'vue'

import { convertCssPixel } from '@idux/cdk/utils'
import { IxIcon } from '@idux/components/icon'

import { DRAWER_TOKEN, drawerToken } from './token'

export default defineComponent({
  setup() {
    const { props, slots } = inject(drawerToken)!
    const { close } = inject(DRAWER_TOKEN)!

    const placement = computed(() => props.placement ?? 'right')
    const classes = computed(() => `ix-drawer ix-drawer-${placement.value}`)
    const isHorizontal = computed(() => placement.value === 'right' || placement.value === 'left')
    const drawerStyle = computed(() => {
      const width = props.width || (isHorizontal.value ? '30%' : '100%') // 不传值时水平方向宽度默认30%
      const height = props.height || (isHorizontal.value ? '100%' : '30%') // 不传值时垂直方向高度默认30%
      const renderOffset = convertCssPixel(props.offset)
      const offset = isHorizontal.value ? { top: renderOffset } : { left: renderOffset }
      return {
        width: convertCssPixel(width),
        height: convertCssPixel(height),
        ...offset,
      }
    })
    return () => {
      const titleElement = renderTitle(slots.title, props.title, props.closable, close)
      const footerElement = renderFooter(slots.footer, props.footer)
      return (
        <div class={classes.value} style={drawerStyle.value}>
          {titleElement}
          <section class="ix-drawer-body">{slots.default?.()}</section>
          {footerElement}
        </div>
      )
    }
  },
})

const renderTitle = (
  titleSlot: Slot | undefined,
  title: string | undefined,
  closable: boolean | undefined,
  close: (evt: Event) => void,
) => {
  if (!titleSlot && !title) {
    return null
  }
  const child = titleSlot ? titleSlot() : <span title={title}>{title}</span>
  return (
    <header class="ix-drawer-header">
      {child}
      {renderIcon(closable, close)}
    </header>
  )
}

const renderIcon = (closable: boolean | undefined, close: (evt: Event) => void) => {
  if (!closable) {
    return null
  }
  return (
    <button class="ix-drawer-close-btn" type="button" onClick={close}>
      <IxIcon class="ix-icon-close" name="close"></IxIcon>
    </button>
  )
}

const renderFooter = (footerSlot: Slot | undefined, footer: string | undefined) => {
  if (!footerSlot && !footer) {
    return null
  }
  const child = footerSlot ? footerSlot() : <span title={footer}>{footer}</span>
  return <footer class="ix-drawer-footer">{child}</footer>
}
