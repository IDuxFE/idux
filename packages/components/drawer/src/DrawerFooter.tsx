/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { VNodeTypes } from 'vue'

import { defineComponent, inject, isVNode } from 'vue'

import { IxButton } from '@idux/components/button'

import { drawerToken } from './token'

export default defineComponent({
  setup() {
    const { props, slots, mergedPrefixCls } = inject(drawerToken)!

    return () => {
      const { footer } = props
      if (!footer && !slots.footer) {
        return null
      }

      let children: VNodeTypes
      if (slots.footer) {
        children = slots.footer()
      } else {
        if (isVNode(footer)) {
          children = footer
        } else {
          children = footer!.map(item => {
            const { text, ...rest } = item
            return <IxButton {...rest}>{text}</IxButton>
          })
        }
      }

      return <div class={`${mergedPrefixCls.value}-footer`}>{children}</div>
    }
  },
})
