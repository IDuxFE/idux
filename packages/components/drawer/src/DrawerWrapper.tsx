/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { DrawerProps } from './types'
import type { DrawerConfig } from '@idux/components/config'
import type { ComputedRef, Ref } from 'vue'

import {
  Transition,
  computed,
  defineComponent,
  inject,
  normalizeClass,
  onBeforeUnmount,
  onMounted,
  ref,
  watch,
} from 'vue'

import { isFunction } from 'lodash-es'

import { callEmit, convertCssPixel } from '@idux/cdk/utils'
import { ɵFooter } from '@idux/components/_private/footer'
import { ɵHeader } from '@idux/components/_private/header'

import { DRAWER_TOKEN, drawerToken } from './token'

const drawerTransitionMap = {
  top: 'move-up',
  bottom: 'move-down',
  start: 'move-start',
  end: 'move-end',
}
const horizontalPlacement = ['start', 'end']
const defaultDistance = 160

export default defineComponent({
  inheritAttrs: false,
  setup(_, { attrs }) {
    const { props, slots, common, config, mergedPrefixCls, visible, animatedVisible, level, levelAction } =
      inject(drawerToken)!
    const { close } = inject(DRAWER_TOKEN)!
    const { closable, closeIcon, closeOnEsc, mask, maskClosable, zIndex } = useConfig(props, config)

    const transitionName = computed(() => `${common.prefixCls}-${drawerTransitionMap[props.placement]}`)
    const isHorizontal = computed(() => horizontalPlacement.includes(props.placement))
    const placementStyle = computed(() => {
      const { width, height, offset, placement } = props
      const horizontal = isHorizontal.value
      const offsetPixel = convertCssPixel(offset)
      let offsetObj
      if (horizontal) {
        const transformPlacement = placement === 'start' ? 'left' : 'right'
        offsetObj = { top: offsetPixel, [transformPlacement]: 0 }
      } else {
        offsetObj = { left: offsetPixel, [placement]: 0 }
      }
      return {
        width: convertCssPixel(width || (horizontal ? config.width : '100%')),
        height: convertCssPixel(height || (horizontal ? '100%' : config.height)),
        ...offsetObj,
      }
    })

    const transformStyle = computed(() => {
      const { placement } = props
      const horizontal = isHorizontal.value
      const distance = level.value * defaultDistance
      let transform
      if (horizontal) {
        transform = distance > 0 ? `translateX(${placement === 'start' ? distance : -distance}px)` : undefined
      } else {
        transform = distance > 0 ? `translateY(${placement === 'top' ? distance : -distance}px)` : undefined
      }
      return transform
    })

    const wrapperClasses = computed(() => {
      const { wrapperClassName = '' } = props
      const action = levelAction.value
      const prefixCls = mergedPrefixCls.value
      return normalizeClass({
        [`${prefixCls}-wrapper`]: true,
        [`${prefixCls}-${props.placement}`]: true,
        [`${prefixCls}-opened`]: animatedVisible.value,
        [`${prefixCls}-${action}`]: !!action,
        [`${prefixCls}-with-mask`]: mask.value,
        [wrapperClassName]: !!wrapperClassName,
      })
    })

    const wrapperStyle = computed(() => {
      const placement = mask.value ? undefined : placementStyle.value
      return { zIndex: zIndex.value, transform: transformStyle.value, ...placement }
    })

    const contentStyle = computed(() => {
      return mask.value ? placementStyle.value : undefined
    })

    const wrapperRef = ref<HTMLDivElement>()
    const sentinelStartRef = ref<HTMLDivElement>()
    const sentinelEndRef = ref<HTMLDivElement>()

    const { onWrapperClick, onWrapperKeydown, onContentMousedown, onContentMouseup } = useEvent(
      close,
      mask,
      maskClosable,
      closeOnEsc,
      sentinelStartRef,
      sentinelEndRef,
    )

    const { onEnter, onAfterEnter, onAfterLeave } = useEvents(props, wrapperRef, animatedVisible)

    onMounted(() => {
      watchVisibleChange(props, wrapperRef, sentinelStartRef, mask)
    })

    return () => {
      const prefixCls = mergedPrefixCls.value
      return (
        <Transition
          name={transitionName.value}
          appear
          onEnter={onEnter}
          onAfterEnter={onAfterEnter}
          onAfterLeave={onAfterLeave}
        >
          <div
            v-show={visible.value}
            ref={wrapperRef}
            class={wrapperClasses.value}
            style={wrapperStyle.value}
            tabindex={-1}
            onClick={onWrapperClick}
            onKeydown={onWrapperKeydown}
          >
            <Transition name={transitionName.value}>
              <div
                role="document"
                class={prefixCls}
                style={contentStyle.value}
                onMousedown={onContentMousedown}
                onMouseup={onContentMouseup}
                {...attrs}
              >
                <div ref={sentinelStartRef} tabindex={0} class={`${prefixCls}-sentinel`} aria-hidden={true}></div>
                <div class={`${prefixCls}-content`}>
                  <ɵHeader
                    v-slots={slots}
                    closable={closable.value}
                    closeIcon={closeIcon.value}
                    header={props.header}
                    onClose={close}
                  />
                  <div class={`${prefixCls}-body`}>{slots.default?.()}</div>
                  <ɵFooter v-slots={slots} class={`${prefixCls}-footer`} footer={props.footer}></ɵFooter>
                </div>
                <div ref={sentinelEndRef} tabindex={0} class={`${prefixCls}-sentinel`} aria-hidden={true}></div>
              </div>
            </Transition>
          </div>
        </Transition>
      )
    }
  },
})

function useConfig(props: DrawerProps, config: DrawerConfig) {
  const closable = computed(() => props.closable ?? config.closable)
  const closeIcon = computed(() => props.closeIcon ?? config.closeIcon)
  const closeOnEsc = computed(() => props.closeOnEsc ?? config.closeOnEsc)
  const mask = computed(() => props.mask ?? config.mask)
  const maskClosable = computed(() => props.maskClosable ?? config.maskClosable)
  const zIndex = computed(() => props.zIndex ?? config.zIndex)

  return { closable, closeIcon, closeOnEsc, mask, maskClosable, zIndex }
}

function watchVisibleChange(
  props: DrawerProps,
  wrapperRef: Ref<HTMLDivElement | undefined>,
  sentinelStartRef: Ref<HTMLDivElement | undefined>,
  mask: ComputedRef<boolean>,
) {
  let lastOutSideActiveElement: HTMLElement | null = null
  watch(
    () => props.visible,
    visible => {
      if (visible) {
        const wrapperElement = wrapperRef.value!
        const activeElement = document.activeElement
        if (!wrapperElement.contains(activeElement)) {
          lastOutSideActiveElement = activeElement as HTMLElement
          sentinelStartRef.value?.focus()
        }
      } else {
        if (mask.value) {
          lastOutSideActiveElement?.focus?.()
          lastOutSideActiveElement = null
        }
      }
    },
    { immediate: true },
  )
}

function useEvent(
  close: (evt: Event) => void,
  mask: ComputedRef<boolean>,
  maskClosable: ComputedRef<boolean>,
  closeOnEsc: ComputedRef<boolean>,
  sentinelStartRef: Ref<HTMLDivElement | undefined>,
  sentinelEndRef: Ref<HTMLDivElement | undefined>,
) {
  let timeId: number | undefined
  let mouseDown = false

  const clearTimer = () => {
    if (timeId) {
      clearTimeout(timeId)
      timeId = undefined
    }
  }

  const onWrapperClick = (evt: MouseEvent) => {
    if (evt.target === evt.currentTarget && !mouseDown && mask.value && maskClosable.value) {
      close(evt)
    }
  }

  const onWrapperKeydown = (evt: KeyboardEvent) => {
    if (closeOnEsc.value && evt.code === 'Escape') {
      evt.stopPropagation()
      close(evt)
      return
    }

    if (evt.code === 'Tab') {
      const activeElement = document.activeElement
      const sentinelStartElement = sentinelStartRef.value
      const sentinelEndElement = sentinelEndRef.value
      if (evt.shiftKey) {
        if (activeElement === sentinelStartElement) {
          sentinelEndElement?.focus()
        }
      } else if (activeElement === sentinelEndElement) {
        sentinelStartElement?.focus()
      }
    }
  }

  const onContentMousedown = () => {
    clearTimer()
    mouseDown = true
  }

  const onContentMouseup = () => {
    if (mouseDown) {
      timeId = setTimeout(() => (mouseDown = false))
    }
  }

  onBeforeUnmount(() => clearTimer())

  return { onContentMousedown, onContentMouseup, onWrapperClick, onWrapperKeydown }
}

function useEvents(
  props: DrawerProps,
  wrapperRef: Ref<HTMLDivElement | undefined>,
  animatedVisible: Ref<boolean | undefined>,
) {
  let lastOutSideActiveElement: HTMLElement | null = null
  const onEnter = () => {
    const wrapperElement = wrapperRef.value!
    const activeElement = document.activeElement
    if (!wrapperElement.contains(activeElement)) {
      lastOutSideActiveElement = activeElement as HTMLElement
    }
  }

  const onAfterEnter = () => {
    const wrapperElement = wrapperRef.value!
    const activeElement = document.activeElement
    if (!wrapperElement.contains(activeElement)) {
      wrapperElement.focus()
    }

    callEmit(props.onAfterOpen)
    animatedVisible.value = true
  }

  const onAfterLeave = () => {
    if (lastOutSideActiveElement && isFunction(lastOutSideActiveElement.focus)) {
      const wrapperElement = wrapperRef.value!
      const activeElement = document.activeElement

      if (
        !activeElement ||
        activeElement === document.body ||
        activeElement === wrapperElement ||
        wrapperElement.contains(activeElement)
      ) {
        lastOutSideActiveElement.focus()
      }
    }

    callEmit(props.onAfterClose)
    animatedVisible.value = false
  }
  return { onEnter, onAfterEnter, onAfterLeave }
}
