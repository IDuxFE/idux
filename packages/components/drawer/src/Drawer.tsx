/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { DrawerProps } from './types'
import type { ComputedRef, Ref } from 'vue'

import { computed, defineComponent, inject, onBeforeUnmount, onMounted, provide, ref, watch, watchEffect } from 'vue'

import { CdkPortal, useTarget } from '@idux/cdk/portal'
import { BlockScrollStrategy } from '@idux/cdk/scroll'
import { callEmit, useControlledProp } from '@idux/cdk/utils'
import { ɵMask } from '@idux/components/_private'
import { useGlobalConfig } from '@idux/components/config'

import DrawerWrapper from './DrawerWrapper'
import { DRAWER_TOKEN, drawerToken } from './token'
import { drawerProps } from './types'

export default defineComponent({
  name: 'IxDrawer',
  inheritAttrs: false,
  props: drawerProps,
  setup(props, { slots, expose, attrs }) {
    const common = useGlobalConfig('common')
    const mergedPrefixCls = computed(() => `${common.prefixCls}-drawer`)
    const config = useGlobalConfig('drawer')
    const mask = computed(() => props.mask ?? config.mask)
    const zIndex = computed(() => props.zIndex ?? config.zIndex)

    const { visible, setVisible, animatedVisible, mergedVisible } = useVisible(props, mask)
    const { open, close } = useTrigger(props, setVisible)
    const { level, levelAction, push, pull } = useLevel(visible)

    provide(drawerToken, {
      props,
      slots,
      common,
      config,
      mergedPrefixCls,
      visible,
      animatedVisible,
      mergedVisible,
      level,
      levelAction,
      push,
      pull,
    })

    const apis = { open, close }
    provide(DRAWER_TOKEN, apis)
    expose(apis)

    return () => {
      if (!mergedVisible.value && props.destroyOnHide) {
        return null
      }
      const { target } = props
      return (
        <CdkPortal target={target ?? `${mergedPrefixCls.value}-container`} load={visible.value}>
          <ɵMask
            class={`${mergedPrefixCls.value}-mask`}
            mask={mask.value}
            visible={visible.value}
            zIndex={zIndex.value}
          />
          <DrawerWrapper {...attrs}></DrawerWrapper>
        </CdkPortal>
      )
    }
  },
})

function useVisible(props: DrawerProps, mask: ComputedRef<boolean>) {
  const [visible, setVisible] = useControlledProp(props, 'visible', false)
  const animatedVisible = ref<boolean>()

  const mergedVisible = computed(() => {
    const currVisible = visible.value
    const currAnimatedVisible = animatedVisible.value
    if (currAnimatedVisible === undefined || currVisible) {
      return currVisible
    }
    return currAnimatedVisible
  })

  let scrollStrategy: BlockScrollStrategy | undefined

  const target = computed(() => {
    const { target } = props
    return target ? useTarget(target) : undefined
  })

  watch(target, value => scrollStrategy?.update({ target: value }))

  watchEffect(() => {
    if (mask.value) {
      if (mergedVisible.value) {
        if (!scrollStrategy) {
          scrollStrategy = new BlockScrollStrategy({ target: target.value })
        }
        scrollStrategy.enable()
      } else {
        scrollStrategy?.disable()
      }
    }
  })

  onBeforeUnmount(() => scrollStrategy?.disable())

  return { visible, setVisible, animatedVisible, mergedVisible }
}

function useTrigger(props: DrawerProps, setVisible: (visible: boolean) => void) {
  const open = () => setVisible(true)

  const close = async (evt?: Event | unknown) => {
    const result = await callEmit(props.onClose, evt)
    if (result === false) {
      return
    }
    setVisible(false)
  }

  return { open, close }
}

function useLevel(visible: Ref<boolean>) {
  const parentContext = inject(drawerToken, null)

  const level = ref(0)
  const levelAction = ref<'push' | 'pull'>()

  const push = () => {
    level.value++
    parentContext?.push()
  }

  const pull = () => {
    level.value--
    parentContext?.pull()
  }

  watch(visible, value => {
    if (value) {
      parentContext?.push()
    } else {
      parentContext?.pull()
      levelAction.value = undefined
    }
  })

  watch(level, (currLevel, prevLevel) => {
    levelAction.value = currLevel > prevLevel ? 'push' : 'pull'
  })

  onMounted(() => {
    if (visible.value) {
      parentContext?.push()
    }
  })

  onBeforeUnmount(() => {
    if (visible.value) {
      parentContext?.pull()
    }
  })

  return { level, levelAction, push, pull }
}
