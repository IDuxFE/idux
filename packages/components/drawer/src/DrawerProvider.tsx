/**
 * @license
 *
 * Use of this source code is governed by an MIT-style license that can be
 * found in the LICENSE file at https://github.com/IDuxFE/idux/blob/main/LICENSE
 */

import type { DrawerInstance, DrawerOptions, DrawerRef } from './types'

import { cloneVNode, defineComponent, isVNode, provide, shallowRef } from 'vue'

import { callEmit, convertArray, noop, uniqueId } from '@idux/cdk/utils'

import Drawer from './Drawer'
import { drawerProviderToken } from './token'

export default defineComponent({
  name: 'IxDrawerProvider',
  setup(_, { expose, slots }) {
    const { drawers, apis } = useDrawerApis()
    provide(drawerProviderToken, apis)
    expose(apis)

    return () => {
      const children = drawers.value.map(item => {
        const { content, contentProps, ...rest } = item
        const contentNode = isVNode(content) ? cloneVNode(content, contentProps, true) : content
        return <Drawer {...rest}>{contentNode}</Drawer>
      })
      return (
        <>
          {slots.default?.()}
          {children}
        </>
      )
    }
  },
})

function useDrawer() {
  const drawers = shallowRef<Array<DrawerOptions & { ref?: (instance: unknown) => void }>>([])
  const drawerRefMap = new Map<string, DrawerRef>()

  const setDrawerRef = (key: string, instance: DrawerInstance | null) => {
    const ref = drawerRefMap.get(key)
    if (instance) {
      if (ref && !ref.open) {
        ref.open = instance.open
        ref.close = instance.close
      }
    } else {
      if (ref) {
        drawerRefMap.delete(key)
        ref.open = noop
        ref.close = noop as (evt?: unknown) => Promise<void>
      }
    }
  }

  const getCurrIndex = (key: string) => {
    return drawers.value.findIndex(message => message.key === key)
  }

  const add = (item: DrawerOptions) => {
    const currIndex = item.key ? getCurrIndex(item.key) : -1
    const tempDrawers = [...drawers.value]
    if (currIndex !== -1) {
      tempDrawers.splice(currIndex, 1, item)
      drawers.value = tempDrawers
      return item.key!
    }
    // The default value for `visible`
    const { key = uniqueId('ix-drawer'), visible = true, destroyOnHide, ...rest } = item
    const setRef = (instance: unknown) => setDrawerRef(key, instance as DrawerInstance | null)
    const onAfterClose = destroyOnHide ? () => destroy(key) : undefined
    tempDrawers.push({ ...rest, key, visible, ref: setRef, onAfterClose })
    drawers.value = tempDrawers
    return key
  }

  const update = (key: string, item: DrawerOptions) => {
    const currIndex = getCurrIndex(key)
    if (currIndex !== -1) {
      const tempDrawers = [...drawers.value]
      const newItem = { ...drawers.value[currIndex], ...item }
      tempDrawers.splice(currIndex, 1, newItem)
      drawers.value = tempDrawers
    }
  }

  const destroy = (key: string | string[]) => {
    const keys = convertArray(key)
    keys.forEach(key => {
      const currIndex = getCurrIndex(key)
      if (currIndex !== -1) {
        const tempDrawers = [...drawers.value]
        const item = tempDrawers.splice(currIndex, 1)
        drawers.value = tempDrawers
        callEmit(item[0].onDestroy, key)
      }
    })
  }

  const destroyAll = () => {
    drawers.value = []
  }

  return { drawers, drawerRefMap, add, update, destroy, destroyAll }
}

function useDrawerApis() {
  const { drawers, drawerRefMap, add, update, destroy, destroyAll } = useDrawer()

  const open = (options: DrawerOptions): DrawerRef => {
    const key = add(options)
    const drawerRef = {
      key,
      update: (options: DrawerOptions) => update(key, options),
      destroy: () => destroy(key),
    } as DrawerRef
    drawerRefMap.set(key, drawerRef)
    return drawerRef
  }

  const apis = { open, update, destroy, destroyAll }

  return { drawers, apis }
}
