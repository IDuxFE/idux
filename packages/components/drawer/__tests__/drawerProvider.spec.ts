import { MountingOptions, VueWrapper, flushPromises, mount } from '@vue/test-utils'
import { h } from 'vue'

import { isElementVisible, renderWork } from '@tests'

import DrawerProvider from '../src/DrawerProvider'
import { DrawerProviderInstance } from '../src/types'

describe('DrawerProvider', () => {
  const DrawerProviderMount = (options?: MountingOptions<Record<string, never>>) => {
    return mount(DrawerProvider, {
      ...options,
    }) as VueWrapper<DrawerProviderInstance>
  }

  const wrapper = DrawerProviderMount()
  let header = 'This is header'
  const content = 'Some contents...'

  afterEach(() => {
    header = 'This is header'
    const container = document.querySelector('.ix-drawer-container')
    if (container) {
      container.innerHTML = ''
    }
  })

  describe('basic', () => {
    renderWork(DrawerProvider, { slots: { default: () => h('p', 'Some contents...') } })

    test('open work', async () => {
      const drawerRef = wrapper.vm.open({ header, content })
      await flushPromises()

      expect(document.querySelector('.ix-header-title')!.textContent).toBe(header)
      expect(document.querySelector('.ix-drawer-body')!.textContent).toBe(content)

      drawerRef.destroy()
    })

    test('update work', async () => {
      const drawerRef = wrapper.vm.open({ header, content })
      await flushPromises()

      expect(document.querySelector('.ix-header-title')!.textContent).toBe(header)
      expect(document.querySelector('.ix-drawer-body')!.textContent).toBe(content)

      header = 'This is header2'
      wrapper.vm.update(drawerRef.key, { header })
      await flushPromises()

      expect(document.querySelector('.ix-header-title')!.textContent).toBe(header)

      drawerRef.destroy()
    })

    test('destroy work', async () => {
      const drawerRef = wrapper.vm.open({ header, content })
      await flushPromises()

      expect(document.querySelectorAll('.ix-drawer-wrapper').length).toBe(1)

      wrapper.vm.destroy(drawerRef.key)
      await flushPromises()

      expect(document.querySelectorAll('.ix-drawer-wrapper').length).toBe(0)
    })

    test('destroyAll work', async () => {
      wrapper.vm.open({ header, content })
      wrapper.vm.open({ header, content })
      wrapper.vm.open({ header, content })

      await flushPromises()

      expect(document.querySelectorAll('.ix-drawer-wrapper').length).toBe(3)

      wrapper.vm.destroyAll()
      await flushPromises()

      expect(document.querySelectorAll('.ix-drawer-wrapper').length).toBe(0)
    })
  })

  describe('drawerRef', () => {
    test('open and close work', async () => {
      const onClose = jest.fn()
      const drawerRef = wrapper.vm.open({ onClose, destroyOnHide: false, header, content })
      await flushPromises()

      drawerRef.close('close')
      await flushPromises()

      expect(onClose).toBeCalledWith('close')
      expect(isElementVisible(document.querySelector('.ix-drawer'))).toBe(false)

      drawerRef.open()
      await flushPromises()

      expect(isElementVisible(document.querySelector('.ix-drawer'))).toBe(true)
    })

    test('destroy work', async () => {
      const onDestroy = jest.fn()
      const drawerRef = wrapper.vm.open({ onDestroy, header, content })
      await flushPromises()

      drawerRef.destroy()
      await flushPromises()

      expect(onDestroy).toBeCalled()
      expect(document.querySelectorAll('.ix-drawer-wrapper').length).toBe(0)
    })
  })
})
