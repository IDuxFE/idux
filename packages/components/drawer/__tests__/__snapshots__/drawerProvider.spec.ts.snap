// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DrawerProvider basic render work 1`] = `"<p>Some contents...</p>"`;
